File: backend/src/models/investorWallet.js
const { query } = require('../db');

function toPublic(row) {
  if (!row) return null;
  return {
    id: row.id,
    investorId: row.investor_id,
    type: row.type,
    amount: Number(row.amount),
    label: row.label,
    status: row.status,
    commitmentId: row.commitment_id || null,
    bankAccountLast4: row.bank_account_last4,
    createdAt: row.created_at,
  };
}

function listByInvestor(investorId) {
  return query(
    `SELECT * FROM investor_wallet_transactions WHERE investor_id = $1 ORDER BY created_at DESC`,
    [investorId]
  ).then((r) => r.rows);
}

/**
 * Available balance = cleared payouts + all withdrawals/reinvestments
 * (withdrawals reserve funds the moment they're requested, before they clear —
 * same pattern as the agent wallet; reinvestments clear immediately).
 * Pending payouts are reported separately and don't count toward the spendable
 * balance yet.
 */
async function getSummary(investorId) {
  const { rows } = await query(
    `SELECT
       COALESCE(SUM(amount) FILTER (WHERE type = 'payout' AND status = 'cleared'), 0)
         + COALESCE(SUM(amount) FILTER (WHERE type IN ('withdrawal', 'reinvestment')), 0) AS balance,
       COALESCE(SUM(amount) FILTER (WHERE type = 'payout' AND status = 'pending'), 0) AS pending_clearance
     FROM investor_wallet_transactions
     WHERE investor_id = $1`,
    [investorId]
  );
  const row = rows[0] || { balance: 0, pending_clearance: 0 };
  return {
    balance: Number(row.balance),
    pendingClearance: Number(row.pending_clearance),
  };
}

function addPayout(investorId, { amount, label, status = 'cleared', commitmentId }) {
  return query(
    `INSERT INTO investor_wallet_transactions (investor_id, type, amount, label, status, commitment_id)
     VALUES ($1, 'payout', $2, $3, $4, $5)
     RETURNING *`,
    [investorId, amount, label, status, commitmentId || null]
  ).then((r) => r.rows[0]);
}

/** Withdrawals are recorded as a negative amount, pending admin clearance. */
function requestWithdrawal(investorId, { amount, bankAccountLast4, label }) {
  return query(
    `INSERT INTO investor_wallet_transactions (investor_id, type, amount, label, status, bank_account_last4)
     VALUES ($1, 'withdrawal', $2, $3, 'pending', $4)
     RETURNING *`,
    [investorId, -Math.abs(amount), label, bankAccountLast4 || null]
  ).then((r) => r.rows[0]);
}

/**
 * Rolls part of the investor's existing wallet balance into a new
 * investment commitment instead of withdrawing it to a bank account.
 * Recorded as a negative amount (same convention as a withdrawal) but
 * clears immediately — the money never leaves the platform, it just
 * moves from spendable balance into a Pending investment_commitments
 * row. See the /api/investors/:id/wallet/reinvest route, which creates
 * the commitment first and passes its id here as commitmentId.
 */
function reinvest(investorId, { amount, label, commitmentId }) {
  return query(
    `INSERT INTO investor_wallet_transactions (investor_id, type, amount, label, status, commitment_id)
     VALUES ($1, 'reinvestment', $2, $3, 'cleared', $4)
     RETURNING *`,
    [investorId, -Math.abs(amount), label, commitmentId || null]
  ).then((r) => r.rows[0]);
}

function clearTransaction(investorId, txId) {
  return query(
    `UPDATE investor_wallet_transactions
     SET status = 'cleared'
     WHERE id = $1 AND investor_id = $2 AND status = 'pending'
     RETURNING *`,
    [txId, investorId]
  ).then((r) => r.rows[0] || null);
}

module.exports = {
  toPublic,
  listByInvestor,
  getSummary,
  addPayout,
  requestWithdrawal,
  reinvest,
  clearTransaction,
};

File: lib/models/investor_wallet.dart
/// A single ledger entry in an investor's wallet — a payout credited by
/// an admin against a confirmed investment commitment, or a withdrawal
/// the investor requested.
class InvestorWalletTransaction {
  final String id;
  final String investorId;
  final String type; // 'payout' | 'withdrawal'
  final double amount;
  final String label;
  final String status; // 'pending' | 'cleared'
  final String? commitmentId;
  final String? bankAccountLast4;
  final DateTime createdAt;

  const InvestorWalletTransaction({
    required this.id,
    required this.investorId,
    required this.type,
    required this.amount,
    required this.label,
    required this.status,
    required this.commitmentId,
    required this.bankAccountLast4,
    required this.createdAt,
  });

  factory InvestorWalletTransaction.fromJson(Map<String, dynamic> json) {
    return InvestorWalletTransaction(
      id: json['id'] as String? ?? '',
      investorId: json['investorId'] as String? ?? '',
      type: json['type'] as String? ?? 'payout',
      amount: (json['amount'] as num?)?.toDouble() ?? 0,
      label: json['label'] as String? ?? '',
      status: json['status'] as String? ?? 'pending',
      commitmentId: json['commitmentId'] as String?,
      bankAccountLast4: json['bankAccountLast4'] as String?,
      createdAt: DateTime.tryParse(json['createdAt'] as String? ?? '') ?? DateTime.now(),
    );
  }
}

/// The balance summary + full ledger returned by GET
/// `/api/investors/:id/wallet`.
class InvestorWalletSummary {
  final double balance;
  final double pendingClearance;
  final List<InvestorWalletTransaction> transactions;

  const InvestorWalletSummary({
    required this.balance,
    required this.pendingClearance,
    required this.transactions,
  });

  factory InvestorWalletSummary.fromJson(Map<String, dynamic> json) {
    return InvestorWalletSummary(
      balance: (json['balance'] as num?)?.toDouble() ?? 0,
      pendingClearance: (json['pendingClearance'] as num?)?.toDouble() ?? 0,
      transactions: (json['transactions'] as List<dynamic>? ?? [])
          .map((e) => InvestorWalletTransaction.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
}
